Merge kde5

ttuegel opened the pull request.

ttuegel commented:
Just waiting for Travis...

vcunat commented on lib/attrsets.nix:
```diff
@@ -78,6 +78,9 @@ rec {
     listToAttrs (fold (n: ys: let v = set.${n}; in if pred n v then [(nameValuePair n v)] ++ ys else ys) [] (attrNames set));
 
+  /* Remove the named attribute from an attribute set. */
+  removeAttr = name: filterAttrs (n: x: n != name);
```
Do you know ``builtins.removeAttrs``? With that, I think we can drop ``removeAttr``, or at least use ``-s`` in the implementation instead.

vcunat commented on pkgs/applications/kde-apps-14.12/manifest.sh:
```diff
@@ -0,0 +1,15 @@
+  # The extra slash at the end of the URL is necessary to stop wget
+  # from recursing over the whole server! (No, it's not a bug.)
+  ../../build-support/autonix/manifest.sh \
+      http://download.kde.org/stable/applications/14.12.0/ \
```
Note: there's http://download.kde.org/stable/applications/14.12.1/ already.

vcunat replied:
And also http://download.kde.org/stable/frameworks/5.6/, when I'm at it :-)

vcunat commented on pkgs/build-support/autonix/generate-collection.nix:
```diff
@@ -0,0 +1,119 @@
+    applyRewriter
+    repackager
+  ]
+
```
I'm just skimming ATM, but can you explain why most of the file is commented-out?

edolstra commented:
What is `autonix`?

vcunat commented:
Yes, at least some in-source-comment description of autonix would be nice. I haven't yet tried to understand what it's supposed to do :-)

ttuegel committed:
add GFDL-1.2 license

ttuegel committed:
clementine: wrap for GST_PLUGINS_PATH

ttuegel committed:
environment: set paths for Qt plugins and QML

ttuegel committed:
libdbusmenu_qt: add Qt 5 version

ttuegel committed:
phonon: add Qt 5 library and backends

ttuegel committed:
gcc-wrapper: parameterize setup-hook

The default setup-hook for gcc-wrapper adds include directories with
-isystem, which upsets the order -I flags are processed. This adds an
alternative setup-hook that only uses -I flags. The build system's
ordering of -I flags is then respected. This is important when different
packages provide includes with the same name, such as building packages
that depend on Qt4 and Qt5.

ttuegel committed:
poppler: add Qt 5 version

ttuegel committed:
quassel: add Qt 5 version

ttuegel committed:
autonix: add Nix library

ttuegel committed:
libarchive-conduit: add foreign dependency

ttuegel committed:
orion: add GTK3 theme

ttuegel committed:
gitignore: ignore autonix tmp files

ttuegel committed:
add kde-frameworks-5.5

ttuegel committed:
add plasma-5.1.2

ttuegel committed:
add kde-applications-14.12.1

ttuegel committed:
add NixOS module for Plasma desktop

ttuegel commented:
OK, I cleaned this up a bit, besides adding documentation.

There is more on this in the documentation, but autonix is a toolkit I created to automate packaging large collections of software, like KDE.

I updated kde-applications to 14.12.1. I am going to wait on updating kde-frameworks to 5.6 because plasma-5.1 requires kde-frameworks-5.5; plasma-5.2 is due at the end of the month and I plan to update kde-frameworks and plasma together then.

Everything builds, except 4 packages from kde-applications. Two of those packages are intended for KDE developers, so I'm not worried about the breakage. The remaining two packages with failing builds have been disabled.

The only remaining runtime bugs that I am aware of are all upstream. That includes: minor graphics glitches with QML2, broken Breeze window decoration (so we use Oxygen instead), and missing polkit agent (fixed upstream in plasma-5.2).

ttuegel committed:
phonon_qt5_backend_vlc: fix src url

ttuegel committed:
kf5, plasma5, kdeApps: add documentation for maintainers

madjar commented:
Since travis won't be able to do it, I built it, and it compiled fine.

edolstra commented on nixos/modules/services/x11/desktop-managers/plasma5.nix:
```diff
@@ -0,0 +1,230 @@
+{
+  options = {
+
+    services.xserver.desktopManager.plasma5 = {
```
Should we really call this `plasma5` rather than `kde5` or `kde`? I understand upstream has some wacky branding going on, but as a KDE user, I wouldn't really think of searching for "plasma" when looking for the option name to enable KDE.

ttuegel replied:
@edolstra That's a good point. I will rename the NixOS module to `kde5`. I want to keep the `kf5`, `plasma5`, and `kdeApps` attributes for the separate collections, but I will add a `kde5 = kf5 // plasma5 // kdeApps` attribute for the combined packages. Does that sound good?

edolstra commented on pkgs/desktops/plasma-5.1/startkde/startkde.in:
```diff
@@ -0,0 +1,361 @@
+#!@bash@/bin/bash -x
+#
+#  NIXOS KDE STARTUP SCRIPT
```
Why is this (very large) script needed? Doesn't upstream provide a startup script? It did in KDE 4 (`${kde_workspace}/bin/startkde`).

ttuegel replied:
@edolstra They still include that script, but it doesn't come close to working for us. I had to edit it so substantially that I thought it made more sense to include our own copy, rather than trying to maintain dozens of patches for a single script.

edolstra commented on pkgs/top-level/all-packages.nix:
```diff
@@ -4057,6 +4059,18 @@ let
       inherit stdenv gcc binutils libc shell name cross;
     });
 
+  wrapGCCStdInc = glibc: baseGCC: (import ../build-support/gcc-wrapper) {
```
What does `StdInc` mean?

edolstra commented on pkgs/build-support/gcc-wrapper/setup-hook-stdinc.sh:
```diff
@@ -0,0 +1,38 @@
+export NIX_CC=@out@
```
What is this script for?

ttuegel replied:
@edolstra This is an alternate setup hook for gcc-wrapper. As I commented over on 51713fb, KDE packages need their includes to be specified with `-I` rather than `-isystem` because the latter changes gcc's path search order for includes. Many of the KDE applications depend on both Qt 4 and Qt 5 (in different components). The CMake build system selects the order of `-I` flags carefully so that each component sees the correct includes, but gcc-wrapper's use of `-isystem` ruins that careful search order. As a result, several KDE packages cannot be compiled with the gcc-wrapper used in the default stdenv.

The meaning of `StdInc` in `wrapGCCStdInc` is "standard include".

geerds replied:
@ttuegel This is a quite good explanation for uncommented code :smile: I think it would be great if you put it in the file

edolstra commented:
Regarding `autonix`: I'm a bit reluctant to have yet another auto-updater in Nixpkgs, since these tend to be used by only one person and make it harder for others to figure out how to update/change a package. We already have so many: gnupdate, upstream-updater, the X.org updater, the KDE 4 updater, etc.

ttuegel commented:
@edolstra There are three options for maintaining the KDE 5 packages:
1. Package everything by hand, no automation for updates,
2. Adopt an existing auto-update solution, or
3. Create a new auto-update system.
I think we both agree that option 1 is not practical for a software collection of this size. I considered option 2, specifically, adopting the KDE 4 updater. However, KDE 5 is packaged somewhat differently, so by the time I made the necessary modifications to the KDE 4 updater, it would have been an entirely different system. I came to the conclusion that trying to adopt any of the other single-purpose updaters would have led to the same end: yet another single-purpose updater.

I wrote autonix because I have the same concern that you do about all these single-purpose updaters. This is not just an auto-updater, it is a library for writing auto-updaters. I think it is powerful enough to replace all our single-purpose updaters, and I hope it eventually will. (Although, I know a few maintainers are ideologically opposed to the use of Haskell anywhere.) It also has features beyond other auto-updaters, namely: dependency resolution. (Dependencies between packages in the same collection are entirely automatic. Dependencies on packages elsewhere in Nixpkgs requires some manual intervention to map the names correctly.)

Having said all that, I'm willing to cast all this aside, if you have a better suggestion about how to maintain the KDE 5 packages.

vcunat commented:
:-D autonix sounds much like http://xkcd.com/927/, but hopefully it will end better (just as nix packaging did in general).

ttuegel commented:
@vcunat I realize that. In my defense, what we have now is not a bunch of competing standards, but just a mess of ad-hoc scripts.

ttuegel committed:
gcc-wrapper: comment on alternate setup hook

We now have an alternate setup hook for gcc-wrapper that uses -I to add
include paths rather than -isystem. The latter flag can change the
search order specified by the build system. For KDE 5 packages, we don't
want that!

ttuegel committed:
add top-level kde5 attribute

ttuegel committed:
kdeApps_14_12.ark: wrap PATH for unzip

ttuegel committed:
kde5: rename plasma5 desktop module

ttuegel committed:
kde5: warn if enabled with kde4

ttuegel committed:
startkde: explain presence of entire script

ttuegel committed:
startkde: don't split script between NixOS module and package

ttuegel commented:
OK, I've renamed the `plasma5` NixOS module to `kde5`. I agree that is the name our users will expect. I also figured out how to warn users if their configuration enables KDE 4 and KDE 5 at the same time (executable names overlap, it's not safe). I believe I have addressed all the other issues in comments. Are there any remaining objections to merging this?

ttuegel committed:
xdg-utils: patch xdg-open to use KDE 5 tools

ttuegel commented:
I just discovered that I did not get the KDE Applications 14.12.1 update merged into this branch, so I'll finish that today.

ttuegel committed:
kdeApps: update to 14.12.1

ttuegel committed:
kdm: install default backgrounds

ttuegel committed:
kde5: simplify package list

ttuegel commented:
OK, that does it for the KDE Applications update. I also fixed an issue with `xdg-open` invoking the wrong `kfmclient`. There is still some weirdness with `kfmclient`, I suspect because it hasn't been ported yet.

ttuegel committed:
add stable attrs for plasma5, kf5, and kdeApps

ttuegel commented:
Just a quick rebase to fix a conflict (XBMC has become Kodi since this branch began). I can't do much more with this without testing from users, this won't change anyone's default settings or replace existing packages, and I've addressed the concerns voiced here, so I'm going to merge it. Anything else, I can fix in master.
